src/Serokell/Util/Group.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE ExplicitForAll      #-}
+{-# LANGUAGE NoImplicitPrelude   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+
+-- | This module introduces functions for grouping.
+
+module Serokell.Util.Group
+       ( groupBy
+       , groupSingleBy
+       ) where
+
+import Universum
+
+import Data.List.NonEmpty ((<|))
+
+import qualified Data.HashMap.Strict as HM
+
+{- | Groups elements using results of the given function as keys.
+
+>>> groupBy even [1,2,3,4,5,6]
+fromList [(False,5 :| [3,1]),(True,6 :| [4,2])]
+
+-}
+groupBy :: forall t a b . (Container t, Element t ~ a, Eq b, Hashable b)
+        => (a -> b) -> t -> HashMap b (NonEmpty a)
+groupBy f = foldl' (hmGroup f) mempty
+  where
+    hmGroup :: (a -> b) -> a -> HashMap b (NonEmpty a) -> HashMap b (NonEmpty a)
+    hmGroup fun x hm =
+        let key = fun x
+            val = case HM.lookup key hm of
+                Nothing -> one x
+                Just xs -> x <| xs
+        in HM.insert key val hm
```

Comment:
`lookup + insert` pattern is already covered by `alter` function:

* http://hackage.haskell.org/package/unordered-containers-0.2.9.0/docs/Data-HashMap-Strict.html#v:alter

You can reuse it here to simplify code a little bit.